courses/advanced-ada/chapters/design_patterns.rst

```diff
@@ -0,0 +1,960 @@
+             else
+                I := I - 1;
+             end if;
+          end;
+
+          if I = 0 then
+             Position.Container := null;
+          end if;
+       end Next;
+
+       procedure Previous (Position : in out Cursor) is
+          I : Natural renames Position.Index;
+
+       begin
+          if I = 0 then
+             return;
+          end if;
+
+          declare
+             S : Stack renames Position.Container.all;
+          begin
+             if I >= S.Top_Index then
+                I := 0;
+                Position.Container := null;
+
+             else
+                I := I + 1;
+             end if;
+          end;
+       end Previous;
+
+    end Stacks.Bounded_G;
+
+.. code:: ada
+
+    with Ada.Unchecked_Deallocation;
+    with System;  use type System.Address;
+
+    package body Stacks.Unbounded_G is
+
+       procedure Free is
+          new Ada.Unchecked_Deallocation
+         (Element_Array,
+          Element_Array_Access);
+
+       procedure Push
+         (Container : in out Stack;
+          Item      :        Element_Type)
+       is
+          R : Rep_Type renames Container.Rep;
+          I : Natural renames R.Top_Index;
+
+       begin
+          if R.Elements = null then
+             R.Elements := new Element_Array'(1 .. 1 => Item);
+             I := 1;
+             return;
+          end if;
+
+          if I = R.Elements'Last then
+             declare
+                X : Element_Array_Access := R.Elements;
+                J : constant Positive := 2 * I;
+                E : Element_Array_Access := new Element_Array (1 .. J);
+
+             begin
+                Copy : begin
+                   E (1 .. I) := X.all;
+                exception
+                   when others =>
+                      Free (E);
+                      raise;
+                end Copy;
+
+                R.Elements := E;
+                Free (X);
+             end;
+          end if;
+
+          R.Elements (I + 1) := Item;
+          I := I + 1;
+       end Push;
+
+       function Top
+         (Container : Stack) return Element_Type
+       is
+          R : Rep_Type renames Container.Rep;
+
+       begin
+          return R.Elements (R.Top_Index);
+       end Top;
+
+       procedure Pop
+         (Container : in out Stack)
+       is
+          R : Rep_Type renames Container.Rep;
+          I : Natural renames R.Top_Index;
+
+       begin
+          I := I - 1;
+       end Pop;
+
+       function Length
+         (Container : Stack) return Natural
+       is (Container.Rep.Top_Index);
+
+       procedure Clear (Container : in out Stack) is
+       begin
+          Container.Rep.Top_Index := 0;
+       end Clear;
+
+       procedure Assign
+         (Target : in out Stack;
+          Source : Stacks.Stack'Class)
+       is
+          --  C : Stacks.Cursor'Class := Source.Top_Cursor;
+          C : Stacks.Cursor'Class := Top_Cursor (Source'Unchecked_Access);
```

Comment:
Ideally, we should remove this `'Unchecked_Access`.

Reply:
Two ideas:

1. Try calling via dot notation: `Source.Top_Cursor`
2. Make Top_Cursor not take an access (I'm not sure why it needs to anyway)

Reply:
Option 1 is actually the commented-out code, so that doesn't work. I suspect option 2 doesn't work either because the access type is used in the implementation of Top_Cursor function (``return Cursor'(Container, R.Top_Index);``). You can find it at line 843.

Reply:
I've just tried the third option (putting the package instantiations into a separate package and used that in the main application). Unfortunately, I still get the same ``accessibility check failed`` error at run time.